### ISSUE TITLE
Ubuntu 24.04.01 returns error for patches

### ISSUE THREAD

**Issue by selcem-artan** [wmcelderry/systemd_with_tpm2#15]
I used cloud-init for generation of ubuntu 22.04.05 image and I am able to systemd-cryptenroll succesfully. So, I tried to apply only patches installation and hooks script. Unfortunately patch install returns error. 
```
uadmin@TRL-U4JF4NY6DC:~/systemd_with_tpm2/patched$      patch cryptsetup_functions ../patches/cryptsetup_functions.patch
patching file cryptsetup_functions
Reversed (or previously applied) patch detected!  Assume -R? [n] 
Apply anyway? [n] y
Hunk #1 FAILED at 84.
Hunk #2 FAILED at 160.
Hunk #3 FAILED at 224.
Hunk #4 succeeded at 365 with fuzz 2 (offset 7 lines).
Hunk #5 FAILED at 382.
4 out of 5 hunks FAILED -- saving rejects to file cryptsetup_functions.rej
```

**Comment by guruevi** [comment]
Here are the changes I made to the various files to get it back to work for Ubuntu 24:

Install tpm2-initramfs-tool and tpm2-tools
Note that tpm2-initramfs-tool is another option to store the password and unlock on boot time, I haven't tried this option yet, since I'm using systemd-cryptenroll to seal the TPM against SecureBoot.
https://github.com/timchen119/tpm2-initramfs-tool

Copy the https://github.com/wmcelderry/systemd_with_tpm2/blob/main/scripts/systemd_cryptsetup_hook
to /etc/initramfs-tools/hooks and chmod a+x

```
diff cryptroot /usr/share/initramfs-tools/scripts/local-top/cryptroot
158c158
<         if [ -z "${CRYPTTAB_OPTION_keyscript+x}" ] && [ "$CRYPTTAB_KEY" != "none" ]; then
---
>         if [ -z "${CRYPTTAB_OPTION_keyscript+x}" ] && ( [ -n "${CRYPTTAB_OPTION_tpm2_device}" ] || [ "$CRYPTTAB_KEY" != "none" ] ); then
```
and
```
diff cryptsetup_functions /usr/lib/cryptsetup/functions
66a67
>            CRYPTTAB_OPTION_tpm2_device \
162a164
>         tpm2-device) OPTION="tpm2_device";; 
226a229,231
>         tpm2-device)
>            [ -n "${VALUE+x}" ] || return 1 # must have a value
>         ;; 
360a366
>     if [[ -z "${CRYPTTAB_OPTION_tpm2_device}" ]] ; then
382a389,391
>     else
>         /lib/systemd/systemd-cryptsetup attach "${CRYPTTAB_NAME}" "${CRYPTTAB_SOURCE}" "${keyfile}" "tpm2-device=${CRYPTTAB_OPTION_tpm2_device}"
>     fi
```
then 
```
systemd-cryptenroll /dev/?da3 --tpm2-device=auto # TPM7 (Secure Boot) is implied
sed -i 's%$%,tpm2-device=auto%' /etc/crypttab
update-initramfs -u
```

**Comment by selcem-artan** [comment]
Hello @guruevi , with 2 diff you shared I understood that you remove tpm2_device related lines from /usr/lib/cryptsetup/functions and  /usr/share/initramfs-tools/scripts/local-top/cryptroot. But that will not fix the TPM enroll issue. Could you please clarify what is changed? I already tried applying 2 patches and cp hook script in 24.04.01 but with reboot TPM does not unseal key.

**Comment by selcem-artan** [comment]
When I apply 2 patches on 24.04 and copy hook script, after reboot TPM can not unseal encryption key. This was working fine in 22.04

![image](https://github.com/user-attachments/assets/2bc9d77f-657b-4980-9ccc-e03c709006c7)

**Comment by guruevi** [comment]
@selcem-artan No, the cryptsetup_functions is the 'original' that ships with 24.04, the /usr/lib/cryptsetup/functions is what I replaced it with, the diff shows what I added and at which lines, you could run these through the patch tool. The cryptroot file indicates a replacement, the rest of the script is identical to what the previous 22.04 version of this repo did.

**Comment by wmcelderry** [comment]
The current code works for me with ubuntu24.0 (well technically the code just before I merged a bunch of stuff, but the diffs haven't changed significantly), so I think this is resolved.  If anyone finds otherwise please comment here, otherwise I'll test it again when I get time.
